jni/src/main/java/com/horizen/storageVersioned/TransactionVersioned.java

```diff
@@ -0,0 +1,143 @@
+    private native void nativeSave() throws Exception;
+    private native void nativeRollbackToSavepoint() throws Exception;
+    private native void nativeRollback() throws Exception;
+    private native DBIterator nativeGetIter(ColumnFamily cf, int mode, byte[] starting_key, int direction) throws Exception;
+    private native ColumnFamily nativeGetColumnFamily(String cf_name) throws Exception;
+
+    // Constructor is intended to be called from inside the Rust environment for setting a raw pointer to a Rust-instance of Transaction
+    private TransactionVersioned(long transactionVersionedPointer, long defaultColumnFamilyPointer) {
+        this.transactionVersionedPointer = transactionVersionedPointer;
+        this.defaultCf = new ColumnFamily(defaultColumnFamilyPointer);
+    }
+
+    // Closes transaction (frees Rust memory from Transaction object)
+    public void closeTransaction() {
+        if (transactionVersionedPointer != 0) {
+            nativeClose(this.transactionVersionedPointer);
+            transactionVersionedPointer = 0;
+        }
+    }
+
+    @Override
+    public void close() {
+        closeTransaction();
+    }
+
+    @Override
+    public ColumnFamily defaultCf() {
+        return defaultCf;
+    }
+
+    public Optional<byte[]> get(ColumnFamily cf, byte[] key){
+        checkPointer();
+        byte[] value = nativeGet(cf, key);
+        if(value != null){
+            return Optional.of(value);
+        } else {
+            return Optional.empty();
+        }
+    }
+
+    public Map<byte[], Optional<byte[]>> get(ColumnFamily cf, Set<byte[]> keys){
+        checkPointer();
+        return nativeMultiGet(cf, keys.toArray(new byte[0][0]));
+    }
+
+    public byte[] getOrElse(ColumnFamily cf, byte[] key, byte[] defaultValue){
+        return get(cf, key).orElse(defaultValue);
+    }
+
+    public boolean isEmpty(ColumnFamily cf) {
+        checkPointer();
+        return nativeIsEmpty(cf);
+    }
+
+    public DBIterator getIter(ColumnFamily cf) throws Exception {
+        // The 'starting_key', and 'direction' parameters are ignored for the 'Start' mode
+        return nativeGetIter(cf, DBIterator.Mode.Start, null, 0);
+    }
+
+    public DBIterator getRIter(ColumnFamily cf) throws Exception {
+        // The 'starting_key', and 'direction' parameters are ignored for the 'End' mode
+        return nativeGetIter(cf, DBIterator.Mode.End, null, 0);
+    }
+
+    public DBIterator getIterFrom(ColumnFamily cf, byte[] starting_key) throws Exception {
+        return nativeGetIter(cf, DBIterator.Mode.From, starting_key, DBIterator.Direction.Forward);
+    }
+
+    public DBIterator getRIterFrom(ColumnFamily cf, byte[] starting_key) throws Exception {
+        return nativeGetIter(cf, DBIterator.Mode.From, starting_key, DBIterator.Direction.Reverse);
+    }
+
+    public void commit(Optional<String> versionIdOpt) throws Exception {
```

Comment:
how can we have an empty version id?
I think, for the versioned transaction/storage we must have next version id.

Reply:
Thanks! Fixed